### DISCUSSION TITLE
New package candidate: Back off

### DISCUSSION THREAD

**Discussion by smebberson** [idearium/idearium-lib#137] [Ideas]
I've recently written and used this within a PR. I couldn't find a nice simple library and so wrote this one.

```JavaScript
'use strict';

const backOffs = [];

module.exports = ({
    name,
    delay = 2000,
    factor = 1.2,
    max = Number.POSITIVE_INFINITY,
}) => {
    if (!backOffs[name]) {
        backOffs[name] = delay;
    }

    return () =>
        Math.min((backOffs[name] += Math.floor(backOffs[name] * factor)), max);
};
```

It's used like this:

```JavaScript
'use strict';

const backOff = require('./back-off');
const inc = backOff({ name: 'uniqueName' });

inc(); // returns 2000
inc(); // returns 2400
inc(); // returns 2880
inc(); // returns 3456
```

You can customize the following within each instance:

- The initial delay.
- The factor (which changes what the back-off path looks like).
- The maximum back-off.

**Comment by allanchau** [comment]
@smebberson I think it looks good. I thought about having max retries instead, but this is probably better since a request won't just stop working after a while.
